cabal-install/Distribution/Client/ProjectPlanning.hs

```diff
@@ -1995,37 +1998,29 @@ fromPrunedPackage (PrunedPackage elab _) = elab
 -- | Set the build targets based on the user targets (but not rev deps yet).
 -- This is required before we can prune anything.
 --
-setRootTargets :: Map UnitId [PackageTarget]
+setRootTargets :: TargetAction
+               -> Map UnitId [ComponentTarget]
                -> [ElaboratedPlanPackage]
                -> [ElaboratedPlanPackage]
-setRootTargets perPkgTargetsMap =
+setRootTargets targetAction perPkgTargetsMap =
     assert (not (Map.null perPkgTargetsMap)) $
     assert (all (not . null) (Map.elems perPkgTargetsMap)) $
 
     map (mapConfiguredPackage setElabBuildTargets)
   where
-    -- Elaborate and set the targets we'll build for this package. This is just
+    -- Set the targets we'll build for this package/component. This is just
     -- based on the root targets from the user, not targets implied by reverse
     -- dependencies. Those comes in the second pass once we know the rev deps.
     --
     setElabBuildTargets elab =
-        elab {
-          elabBuildTargets   = mapMaybe targetForElab buildTargets,
-          elabReplTarget     = replTarget >>= targetForElab,
-          elabBuildHaddocks  = buildHaddocks
-        }
-      where
-        (buildTargets, replTarget, buildHaddocks)
-                = elaboratePackageTargets elab targets
-        targets = fromMaybe []
-                $ Map.lookup (installedUnitId elab) perPkgTargetsMap
-        targetForElab tgt@(ComponentTarget cname _) =
-            case elabPkgOrComp elab of
-                ElabPackage _ -> Just tgt -- always valid
-                ElabComponent comp
-                    -- Only if the component name matches
-                    | compComponentName comp == Just cname -> Just tgt
-                    | otherwise -> Nothing
+      case (Map.lookup (installedUnitId elab) perPkgTargetsMap,
+            targetAction) of
+        (Nothing, _)                      -> elab
+        (Just tgts,  TargetActionBuild)   -> elab { elabBuildTargets = tgts }
+        (Just [tgt], TargetActionRepl)    -> elab { elabReplTarget = Just tgt }
+        (Just _,     TargetActionHaddock) -> elab { elabBuildHaddocks = True }
+        (Just _,     TargetActionRepl)    ->
+          error "pruneInstallPlanToTargets: multiple repl targets"
```

Comment:
I think that was deliberate. `pruneInstallPlanToTargets` is the only exported function that uses it.